panel/src/components/Forms/FormButtons.vue

```diff
@@ -27,24 +62,140 @@ export default {
     hasChanges() {
       return this.$store.getters["form/hasChanges"](this.id);
     },
+    hasLock() {
+      return this.form.lock !== null;
+    },
+    hasUnlock() {
+      return this.form.unlock !== null;
+    },
+    form() {
+      return {
+        lock: this.$store.getters["form/lock"],
+        unlock: this.$store.getters["form/unlock"]
+      };
+    },
     id() {
-      return this.$store.state.form.current;
+      return this.$store.getters["form/current"];
+    },
+    isDisabled() {
+      return this.$store.getters["form/isDisabled"];
+    }
+  },
+  watch: {
+    hasChanges(current, previous) {
+      // if user started to make changes,
+      // start setting lock on each heartbeat
+      if (current === true && previous == false) {
+        this.$store.dispatch("heartbeat/remove", this.listen);
+        this.$store.dispatch("heartbeat/add", [this.setLock, 30]);
+
+      // if user reversed changes manually,
+      // remove lock and listen to lock from other users again
+      } else if (this.id && current === false && previous == true) {
+        this.removeLock();
+      }
     },
-    isLocked() {
-      return this.$store.getters["form/isLocked"];
+    id() {
+      // whenever the model id changes,
+      // make sure to remove heartbeats
+      if (!this.id) {
+        this.$store.dispatch("heartbeat/remove", this.listen);
+        this.$store.dispatch("heartbeat/remove", this.setLock);
+        return;
+      }
+
+      if (this.hasChanges === false) {
+        this.$store.dispatch("heartbeat/add", this.listen);
+      }
     }
   },
   created() {
-    this.$events.$on("keydown.cmd.s", this.save);
+    this.$events.$on("keydown.cmd.s", this.onSave);
   },
   destroyed() {
-    this.$events.$off("keydown.cmd.s", this.save);
+    this.$events.$off("keydown.cmd.s", this.onSave);
   },
   methods: {
-    reset() {
+    listen() {
+      return this.$api.get(this.$route.path + "/lock", null, null, true).then(response => {
+
+        // if content is locked, dispatch info to store
+        if (response.locked === true) {
+          this.$store.dispatch("form/lock", {
+            user: response.user,
+            email: response.email,
+            time: parseInt(response.time, 10),
+            canUnlock: response.canUnlock
+          });
+          return;
+        }
+
+        // if content is not locked but store still holds a lock
+        // from a another user, that lock has been lifted and thus
+        // the content needs to be reloaded to reflect any changes
+        if (
+          this.hasLock &&
+          this.form.lock.user !== this.$store.state.user.current.id
+        ) {
+          this.$events.$emit("model.reload");
+        }
+
+        this.$store.dispatch("form/lock", null);
+      });
+    },
+    setLock() {
+      this.$api.patch(this.$route.path + "/lock", null, null, true).catch(() => {
+        this.$store.dispatch("form/revert", this.id);
```

Comment:
This seems to trigger a request to the `DELETE /lock` API endpoint, which results in another Exception ("The content lock can only be removed by the user who created it.") in the browser console. Does not break the behavior, only a cosmetic issue. If you can fix it without too much effort, that would be great. Otherwise leave it like this. :)

Reply:
I don't think there is an easy way out of this... It's triggered here: https://github.com/getkirby/kirby/pull/1718/files#diff-53f37eaee7ac16a1f56969f2cd5b1a56R95
We need that part, so that a lock gets removed if the user makes a change (setting the lock) and then manually changes the value back to the original. The Panel recognises this as no changes anymore and so the lock should also be removed. But it is also triggered here when reverting the changes directly when discovered that the page is already locked.

Option would be to remove that revert line but then user B would see his/her few changes still in the disabled form inputs (disabled because of the lock by user A), even though they will be ignored immediately when user A saves or reverts.

Reply:
That's what I suspected. The line definitely has its right to exist, I thought of a check like "only try to remove the lock if there was one before". But if that's not doable, feel free to resolve this conversation. Doesn't matter that much.